### PR TITLE
swift4: Minor fixes

### PR DESCRIPTION
**Pull request by spevans**
- Fix 'Cannot call value of non-function type 'UnsafePointer<Int8>'
  by fully qualifying type(of:) in NSValue.swift

Compatible with swift 3 & 4

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

File: Foundation/NSURLResponse.swift [hunk 1]
```diff
@@ -396,8 +396,8 @@ private extension String {
                 type = string
             } else {
                 if let r = string.range(of: "=") {
-                    let name = string[string.startIndex..<r.lowerBound].trimmingCharacters(in: ws)
-                    let value = string[r.upperBound..<string.endIndex].trimmingCharacters(in: ws)
+                    let name = String(string[string.startIndex..<r.lowerBound]).trimmingCharacters(in: ws)
```
Comment (phausler) [inline Foundation/NSURLResponse.swift]: This is fairly unfortunate that we have to force a copy here, In the grand scheme of things it probably won't matter too much since the range is probably always small

File: Foundation/NSValue.swift [hunk 2]
```diff
@@ -98,7 +98,7 @@ open class NSValue : NSObject, NSCopying, NSSecureCoding, NSCoding {
     }
     
     public convenience required init(bytes value: UnsafeRawPointer, objCType type: UnsafePointer<Int8>) {
-        if type(of: self) == NSValue.self {
+        if Swift.type(of: self) == NSValue.self {
```
Comment (phausler) [inline Foundation/NSValue.swift]: imho this seems like a bug of type(of:) to require this! what is the causality here?

File: Foundation/Process.swift [hunk 3]
```diff
@@ -241,7 +241,11 @@ open class Process: NSObject {
         if let env = environment {
             let nenv = env.count
             envp = UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>.allocate(capacity: 1 + nenv)
-            envp.initialize(from: env.map { strdup("\($0)=\($1)") })
+            let evars = env.map { (arg) -> UnsafeMutablePointer<Int8>? in
+                let (key, value) = arg
+                return UnsafeMutablePointer(strdup("\(key)=\(value)"))
+            }
+            envp.initialize(from: evars, count: 1 + nenv)
```
Comment (phausler) [inline Foundation/Process.swift]: hmm there is a size mismatch, nenv is the count of elements in env, so the initialization will be reading past the mapped value.

The expectation here is the envp has a "null terminator"

File: Foundation/NSValue.swift (hunk 2 above)
Comment (spevans) [inline Foundation/NSValue.swift]: See https://lists.swift.org/pipermail/swift-users/Week-of-Mon-20170515/005442.html

Its due to the clash of the 'type' argument and the 'type' function call

Comment (phausler) [inline Foundation/NSValue.swift]: Hmmph that still seems like a bug to me... type as the variable seems quite clear that it is not invoked as a function especially since it has a named parameter. But ok, I guess chock it up to the same failure as print and NSView.print :/

File: Foundation/Process.swift (hunk 3 above)
Comment (spevans) [inline Foundation/Process.swift]: I was actually going to revert this change since there was some discussion on the closure variables that @parkera  raised (the let (key, value) = arg change).

Isnt the mapped size correct since its allocating 1 + env, but the actual issue is we need to ensure its set to NULL to terminate the **env array?

**Comment by spevans** [comment]
@parkera Could you also review these 2 fixes required for swift4 please

**Comment by alblue** [comment]
@swift-ci please test

**Comment by parkera** [comment]
These seem good to me.

**Comment by spevans** [comment]
I think the test got broken due to github's API issues yesterday could someone retest please

**Comment by pushkarnk** [comment]
Seems ready for merge

@swift-ci please test and merge

**Comment by pushkarnk** [comment]
@swift-ci please test and merge
